#Final Project - Data Gathering and Warehousing - DSSA-5102#
Instructor: Melissa Laurino
Spring 2025

##Data Cleaning##
Let's load our data, clean it up and prepare it for use in our final project.

In [ ]:
!pip install python-dotenv

In [ ]:
# Load necessary packages:
import pandas as pd
import numpy as np
import os
from google.colab import drive
from dotenv import load_dotenv

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Path to colab secrets file
secrets_path = '/content/drive/MyDrive/colab_secrets/github.env'
# Load the secrets env file
load_dotenv(secrets_path)
username = os.getenv('GITHUB_USERNAME')
token = os.getenv('GITHUB_TOKEN')
# Clone the repo using credentials
repo_owner = 'Decabra'
repo_name = 'pet-puffs'
repo_url = f"https://{username}:{token}@github.com/{repo_owner}/{repo_name}.git"
# Remove old repo if exists any and loads the latest version from GitHub
!rm -rf $repo_name
!git clone $repo_url

print(repo_url)
print(token)

Cloning into 'pet-puffs'...
remote: Enumerating objects: 45, done.
remote: Counting objects: 100% (45/45), done.
remote: Compressing objects: 100% (34/34), done.
remote: Total 45 (delta 17), reused 16 (delta 6), pack-reused 0 (from 0)
Receiving objects: 100% (45/45), 4.94 MiB | 4.10 MiB/s, done.
Resolving deltas: 100% (17/17), done.
https://joedag32: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@github.com/Decabra/pet-puffs.git
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


In [ ]:
# load AKC csv file
akc_dogs_df = pd.read_csv('/content/pet-puffs/datasets/akc-data-latest.csv')
# load Pet Owners file
pet_owners_df = pd.read_csv('/content/pet-puffs/datasets/pet-owners.csv')

# Breed Dataset

Let's clean up the breed dataset

In [ ]:
# rename the first column to breed
akc_dogs_df.columns.values[0] = "breed"
akc_dogs_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 21 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   breed                        277 non-null    object 
 1   description                  277 non-null    object 
 2   temperament                  276 non-null    object 
 3   popularity                   198 non-null    object 
 4   min_height                   277 non-null    float64
 5   max_height                   277 non-null    float64
 6   min_weight                   275 non-null    float64
 7   max_weight                   275 non-null    float64
 8   min_expectancy               274 non-null    float64
 9   max_expectancy               274 non-null    float64
 10  group                        277 non-null    object 
 11  grooming_frequency_value     270 non-null    float64
 12  grooming_frequency_category  270 non-null    object 
 13  shedding_value      

In [ ]:
# Drop the columns we will not be using
akc_dogs_df = akc_dogs_df.drop(columns=['description', 'group', 'grooming_frequency_value', 'grooming_frequency_category', 'shedding_value', 'shedding_category'])

In [ ]:
# Preview our dataframe after the column drops
akc_dogs_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   breed                  277 non-null    object 
 1   temperament            276 non-null    object 
 2   popularity             198 non-null    object 
 3   min_height             277 non-null    float64
 4   max_height             277 non-null    float64
 5   min_weight             275 non-null    float64
 6   max_weight             275 non-null    float64
 7   min_expectancy         274 non-null    float64
 8   max_expectancy         274 non-null    float64
 9   energy_level_value     271 non-null    float64
 10  energy_level_category  271 non-null    object 
 11  trainability_value     253 non-null    float64
 12  trainability_category  253 non-null    object 
 13  demeanor_value         252 non-null    float64
 14  demeanor_category      252 non-null    object 
dtypes: flo

In [ ]:
# let's look for any rows with missing values
rows_with_missing_values = akc_dogs_df[akc_dogs_df.isna().any(axis=1)]
print(rows_with_missing_values)

                         breed                           temperament  \
5             American Bulldog                 Loyal, Self-Confident   
9    American Hairless Terrier             Energetic, Alert, Curious   
10      American Leopard Hound      Sociable, Energetic, Intelligent   
14      Appenzeller Sennenhund              Agile, Versatile, Lively   
16           Australian Kelpie             Loyal, Alert, Intelligent   
..                         ...                                   ...   
260        Transylvanian Hound  Courageous, Good-Natured, Determined   
261  Treeing Tennessee Brindle          Friendly, Alert, Intelligent   
268                 Wetterhoun      Loyal, Good-Natured, Intelligent   
273             Working Kelpie             Alert, Eager, Intelligent   
275             Yakutian Laika     Affectionate, Intelligent, Active   

    popularity  min_height  max_height  min_weight  max_weight  \
5          NaN       50.80       63.50   27.215542   45.359237   
9  

In [ ]:
# Replace the NaN values in object type columns with None, that seems to work better with MySQL
# set the object type columns
cols_to_replace = ['energy_level_category', 'trainability_category', 'demeanor_category']

# Replace NaN with None in the specified columns
akc_dogs_df[cols_to_replace] = akc_dogs_df[cols_to_replace].where(pd.notna(akc_dogs_df[cols_to_replace]), None)

In [ ]:
akc_dogs_df['popularity'].unique()

array(['148', '113', '60', '47', '58', nan, '175', '122', '186', '136',
       '85', '166', '90', '55', '17', '140', 'of', '87', '39', '6', '127',
       '124', '141', '43', '125', '106', '187', '144', '22', '46', '138',
       '118', '49', '130', '121', '35', '88', '103', '21', '84', '11',
       '100', '132', '26', '98', '62', '5', '51', '69', '179', '32', '68',
       '18', '185', '45', '33', '79', '64', '190', '75', '183', '143',
       '30', '38', '81', '162', '12', '56', '176', '67', '52', '188',
       '94', '27', '135', '157', '149', '161', '184', '91', '4', '134',
       '2', '9', '63', '78', '174', '3', '115', '177', '16', '66', '74',
       '145', '189', '24', '152', '155', '146', '77', '116', '159', '76',
       '73', '104', '95', '129', '173', '163', '1', '99', '147', '93',
       '71', '168', '37', '133', '29', '34', '110', '70', '19', '102',
       '150', '40', '126', '165', '97', '191', '108', '83', '72', '182',
       '54', '117', '92', '13', '156', '172', '171', '114'

In [ ]:
# Convert popularity column to numeric, then replace NaN in the popularity column with 0
akc_dogs_df['popularity'] = pd.to_numeric(akc_dogs_df['popularity'], errors='coerce')
akc_dogs_df['popularity'].fillna(0, inplace=True)

<ipython-input-12-6c50d749ab99>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  akc_dogs_df['popularity'].fillna(0, inplace=True)


In [ ]:
akc_dogs_df['popularity'].unique()
#akc_dogs_df.info()

array([148., 113.,  60.,  47.,  58.,   0., 175., 122., 186., 136.,  85.,
       166.,  90.,  55.,  17., 140.,  87.,  39.,   6., 127., 124., 141.,
        43., 125., 106., 187., 144.,  22.,  46., 138., 118.,  49., 130.,
       121.,  35.,  88., 103.,  21.,  84.,  11., 100., 132.,  26.,  98.,
        62.,   5.,  51.,  69., 179.,  32.,  68.,  18., 185.,  45.,  33.,
        79.,  64., 190.,  75., 183., 143.,  30.,  38.,  81., 162.,  12.,
        56., 176.,  67.,  52., 188.,  94.,  27., 135., 157., 149., 161.,
       184.,  91.,   4., 134.,   2.,   9.,  63.,  78., 174.,   3., 115.,
       177.,  16.,  66.,  74., 145., 189.,  24., 152., 155., 146.,  77.,
       116., 159.,  76.,  73., 104.,  95., 129., 173., 163.,   1.,  99.,
       147.,  93.,  71., 168.,  37., 133.,  29.,  34., 110.,  70.,  19.,
       102., 150.,  40., 126., 165.,  97., 191., 108.,  83.,  72., 182.,
        54., 117.,  92.,  13., 156., 172., 171., 114., 170.,  23.,   7.,
       154.,  50.,  28., 160., 151., 181.,  86., 14

In [ ]:
# let's see the max values in max_height, max_weight, and max_expectancy columns to get an idea of the data range in them
maxAkcValues = akc_dogs_df[['max_height', 'max_weight', 'max_expectancy']].max()
maxAkcValues

max_height         88.900000
max_weight        108.862169
max_expectancy     19.000000
dtype: float64

In [ ]:
# Export the akc_dogs_df dataset to a .csv file
akc_dogs_df.to_csv('pet-puffs/datasets/dog_breed_clean.csv', index=False, sep=',', encoding='utf-8')

# Pet Dataset

Let's clean up the pet dataset

In [ ]:
# preview of the pet owners data
pet_owners_df.head()

Animal ID       Name                DateTime MonthYear Date of Birth  \
0   A882831  *Hamilton  07/01/2023 06:12:00 PM  Jul 2023    03/25/2023   
1   A794011      Chunk  05/08/2019 06:20:00 PM  May 2019    05/02/2017   
2   A776359      Gizmo  07/18/2018 04:02:00 PM  Jul 2018    07/12/2017   
3   A821648        NaN  08/16/2020 11:38:00 AM  Aug 2020    08/16/2019   
4   A720371      Moose  02/13/2016 05:59:00 PM  Feb 2016    10/08/2015   

  Outcome Type Outcome Subtype Animal Type Sex upon Outcome Age upon Outcome  \
0     Adoption             NaN         Cat    Neutered Male         3 months   
1    Rto-Adopt             NaN         Cat    Neutered Male          2 years   
2     Adoption             NaN         Dog    Neutered Male           1 year   
3   Euthanasia             NaN       Other          Unknown           1 year   
4     Adoption             NaN         Dog    Neutered Male         4 months   

                                Breed              Color  
0              Domestic Shorthair Mix        Black/White  
1              Domestic Shorthair Mix  Brown Tabby/White  
2             Chihuahua Shorthair Mix        White/Brown  
3                             Raccoon               Gray  
4  Anatol Shepherd/Labrador Retriever               Buff

In [ ]:
print(pet_owners_df)

       Animal ID       Name                DateTime MonthYear Date of Birth  \
0        A882831  *Hamilton  07/01/2023 06:12:00 PM  Jul 2023    03/25/2023   
1        A794011      Chunk  05/08/2019 06:20:00 PM  May 2019    05/02/2017   
2        A776359      Gizmo  07/18/2018 04:02:00 PM  Jul 2018    07/12/2017   
3        A821648        NaN  08/16/2020 11:38:00 AM  Aug 2020    08/16/2019   
4        A720371      Moose  02/13/2016 05:59:00 PM  Feb 2016    10/08/2015   
...          ...        ...                     ...       ...           ...   
172595   A926855   *Eleanor  03/24/2025 04:30:00 PM  Mar 2025    11/18/2024   
172596   A926900     *Ponzu  03/24/2025 05:34:00 PM  Mar 2025    03/18/2023   
172597   A925467    *Oliver  03/24/2025 05:24:00 PM  Mar 2025    02/24/2023   
172598   A926030    Kashmir  03/24/2025 05:49:00 PM  Mar 2025    03/04/2024   
172599   A926980        NaN  03/24/2025 05:44:00 PM  Mar 2025    01/13/2025   

       Outcome Type Outcome Subtype Animal Type Sex

In [ ]:
# Number of AKC dog breeds
len(akc_dogs_df['breed'].unique())

277

In [ ]:
# Filter pet_owners_df to be only Animal Type of Dog
dog_pet_owners_df = pet_owners_df[pet_owners_df['Animal Type'] == 'Dog']

In [ ]:
# Get an idea of how what kind of unique breed values are in pet
dog_pet_owners_df['Breed'].unique()

array(['Chihuahua Shorthair Mix', 'Anatol Shepherd/Labrador Retriever',
       'American Foxhound/Labrador Retriever', ..., 'Pit Bull/Bloodhound',
       'Shih Tzu/Cairn Terrier', 'French Bulldog/Pomeranian'],
      dtype=object)

In [ ]:
# Assign all the dog breeds that are of mix to pet_contains_slash_df
pet_contains_slash_df = dog_pet_owners_df[dog_pet_owners_df['Breed'].str.contains('Mix', na=False)]
pet_contains_slash_df

Animal ID         Name                DateTime MonthYear Date of Birth  \
2        A776359        Gizmo  07/18/2018 04:02:00 PM  Jul 2018    07/12/2017   
6        A659412     Princess  10/05/2020 02:37:00 PM  Oct 2020    03/24/2013   
17       A843327        *Mary  10/08/2021 01:25:00 PM  Oct 2021    09/29/2019   
20       A595061        Gypsy  08/18/2022 06:07:00 PM  Aug 2022    10/01/2010   
27       A856853      Jameson  05/20/2022 03:32:00 PM  May 2022    11/09/2021   
...          ...          ...                     ...       ...           ...   
172565   A737625  *Lt Winters  03/23/2025 06:06:00 PM  Mar 2025    05/01/2015   
172574   A856897       Junior  05/12/2022 04:01:00 PM  May 2022    03/09/2022   
172586   A926712  *Rocket Man  03/24/2025 04:19:00 PM  Mar 2025    09/15/2024   
172598   A926030      Kashmir  03/24/2025 05:49:00 PM  Mar 2025    03/04/2024   
172599   A926980          NaN  03/24/2025 05:44:00 PM  Mar 2025    01/13/2025   

           Outcome Type Outcome Subtype Animal Type Sex upon Outcome  \
2              Adoption             NaN         Dog    Neutered Male   
6              Adoption             NaN         Dog    Spayed Female   
17             Transfer       Out State         Dog    Intact Female   
20      Return to Owner             NaN         Dog    Spayed Female   
27             Adoption             NaN         Dog    Neutered Male   
...                 ...             ...         ...              ...   
172565         Adoption             NaN         Dog    Neutered Male   
172574         Adoption             NaN         Dog    Neutered Male   
172586         Adoption             NaN         Dog    Neutered Male   
172598         Adoption             NaN         Dog    Spayed Female   
172599         Adoption             NaN         Dog    Neutered Male   

       Age upon Outcome                    Breed        Color  
2                1 year  Chihuahua Shorthair Mix  White/Brown  
6               7 years  Chihuahua Shorthair Mix        Brown  
17              2 years  Chihuahua Shorthair Mix  Black/White  
20             11 years   Labrador Retriever Mix    Red/White  
27             6 months      Black Mouth Cur Mix  Brown/Black  
...                 ...                      ...          ...  
172565          9 years             Pit Bull Mix        White  
172574         2 months      German Shepherd Mix     Tricolor  
172586         6 months         Carolina Dog Mix    Red/White  
172598           1 year   Labrador Retriever Mix  Black/White  
172599         2 months             Pit Bull Mix        Brown  

[59571 rows x 12 columns]

59571 rows cotain " Mix" with the dog breed. We're going to remove " Mix" if the aditional breed wasn't provided with the dataset.

In [ ]:
# Strip " Mix" from the Breed column
dog_pet_owners_df.loc[:, 'Breed'] = dog_pet_owners_df['Breed'].str.rstrip(' Mix')

In [ ]:
# See how many breed matches are between the akc and pets datasets
breed_matches_df = dog_pet_owners_df['Breed'].isin(akc_dogs_df['breed']).sum()
print(f"Toal Pet Owners: {len(dog_pet_owners_df)}")
print(f"Total Pet Owners with AKC Breed Matches: {breed_matches_df}")

Toal Pet Owners: 93978
Total Pet Owners with AKC Breed Matches: 38879


In [ ]:
dog_pet_owners_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 93978 entries, 2 to 172599
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Animal ID         93978 non-null  object
 1   Name              79208 non-null  object
 2   DateTime          93978 non-null  object
 3   MonthYear         93978 non-null  object
 4   Date of Birth     93978 non-null  object
 5   Outcome Type      93951 non-null  object
 6   Outcome Subtype   31146 non-null  object
 7   Animal Type       93978 non-null  object
 8   Sex upon Outcome  93976 non-null  object
 9   Age upon Outcome  93974 non-null  object
 10  Breed             93978 non-null  object
 11  Color             93978 non-null  object
dtypes: object(12)
memory usage: 9.3+ MB


In [ ]:
# See the unique values of Outcome Type column
dog_pet_owners_df['Outcome Type'].unique()

array(['Adoption', 'Return to Owner', 'Transfer', 'Rto-Adopt',
       'Euthanasia', 'Died', 'Missing', nan, 'Disposal', 'Stolen', 'Lost',
       'Relocate'], dtype=object)

In [ ]:
# Create pet_trim_df dataframe with just the columns we want from dog_pet_owners_df
pet_trim_df = dog_pet_owners_df.loc[:, ['Name', 'Date of Birth', 'Outcome Type', 'Outcome Subtype', 'DateTime', 'Sex upon Outcome', 'Color', 'Age upon Outcome', 'Breed']]

In [ ]:
# Rename the columns to be all lowercase with no spaces
pet_trim_df.rename(columns={'Name': 'name', 'Date of Birth': 'dob', 'Outcome Type': 'outcome_type', 'Outcome Subtype': 'outcome_subtype', 'DateTime': 'outcome_datetime', 'Sex upon Outcome': 'sex', 'Color': 'color', 'Age upon Outcome': 'age', 'Breed': 'breed'}, inplace=True)
pet_trim_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 93978 entries, 2 to 172599
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   name              79208 non-null  object
 1   dob               93978 non-null  object
 2   outcome_type      93951 non-null  object
 3   outcome_subtype   31146 non-null  object
 4   outcome_datetime  93978 non-null  object
 5   sex               93976 non-null  object
 6   color             93978 non-null  object
 7   age               93974 non-null  object
 8   breed             93978 non-null  object
dtypes: object(9)
memory usage: 7.2+ MB


In [ ]:
# let's look for any rows with missing values
pet_rows_with_missing_values = pet_trim_df[pet_trim_df.isna().any(axis=1)]
print(pet_rows_with_missing_values)

               name         dob     outcome_type outcome_subtype  \
2             Gizmo  07/12/2017         Adoption             NaN   
4             Moose  10/08/2015         Adoption             NaN   
6          Princess  03/24/2013         Adoption             NaN   
12            Tulip  08/06/2019         Adoption             NaN   
16            Fiona  06/01/2018  Return to Owner             NaN   
...             ...         ...              ...             ...   
172574       Junior  03/09/2022         Adoption             NaN   
172586  *Rocket Man  09/15/2024         Adoption             NaN   
172591   *Charlotte  12/02/2024         Adoption             NaN   
172598      Kashmir  03/04/2024         Adoption             NaN   
172599          NaN  01/13/2025         Adoption             NaN   

              outcome_datetime            sex        color       age  \
2       07/18/2018 04:02:00 PM  Neutered Male  White/Brown    1 year   
4       02/13/2016 05:59:00 PM  Neutere

In [ ]:
# Let's replace the NaN values for the name with Unknown
pet_trim_df['name'] = pet_trim_df['name'].replace(np.nan, 'Unknown')

In [ ]:
# Let's replace the NaN values for the outcome_type with None
pet_trim_df['outcome_type'].fillna('None', inplace=True)

<ipython-input-29-fff63b887672>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  pet_trim_df['outcome_type'].fillna('None', inplace=True)


In [ ]:
# Drop any remaining rows with NaN values
pet_trim_df.dropna(inplace=True)

In [ ]:
# Convert dob to MySQL friendly date format
pet_trim_df['dob'] = pd.to_datetime(pet_trim_df['dob'])
pet_trim_df['dob'] = pet_trim_df['dob'].dt.strftime('%Y-%m-%d')

In [ ]:
# Convert outcome_datetime to MySQL friendly date/time format
pet_trim_df['outcome_datetime'] = pd.to_datetime(pet_trim_df['outcome_datetime'])
pet_trim_df['outcome_datetime'] = pet_trim_df['outcome_datetime'].dt.strftime('%Y-%m-%d %H:%M:%S')

In [ ]:
#Let's keep only rows where pet breed matches the akc_dogs breed
akc_breeds = akc_dogs_df['breed'].unique()
filtered_pet_trim_df = pet_trim_df[pet_trim_df['breed'].isin(akc_breeds)]
len(filtered_pet_trim_df) # output number of clean breed rows

13125

In [ ]:
# Export the filtered_pet_trim_df dataset to a .csv file
filtered_pet_trim_df.to_csv('pet-puffs/datasets/pet_clean.csv', index=False, sep=',', encoding='utf-8')

In [ ]:
# Let's try connecting to a MySQL database hosted on a webserver

#install mysql.connector
!pip install mysql-connector-python mysql-connector

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.9/11.9 MB 58.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.0/34.0 MB 14.9 MB/s eta 0:00:00
  Created wheel for mysql-connector: filename=mysql_connector-2.2.9-cp311-cp311-linux_x86_64.whl size=247951 sha256=2d9360b5f8a96996b3ee002ee1a2e3ff1e0a0e1fa40db23db3b2565ee3e68ab6
  Stored in directory: /root/.cache/pip/wheels/17/cd/ed/2d49e9bac69cf09382e4c7cc20a2511202b48324b87db26019
Successfully built mysql-connector


In [ ]:
# Load necessary packages:
from sqlalchemy import create_engine, Column, String, Integer, Boolean, BigInteger, Float, text # Database navigation
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from sqlalchemy import insert
import mysql.connector

In [ ]:
# Load the Google Colab Secrets for our db pwd
from google.colab import userdata

In [ ]:
# connect to MySQL server

# conn = mysql.connector.connect(
#         host="mysql.joedag32.com",
#         user="petpuffs",
#         password=userdata.get('db_pwd'),
#         database="data_warehousing_final")

# mycursor = conn.cursor()
# mycursor.execute("SELECT VERSION()")
# data = mycursor.fetchone()
# print(f"Database version: {data[0]}")

In [ ]:
#Close the database connection :)
# mycursor.close()
# conn.close()

# **Let's test out a SSH Tunnel** to connect to MySQL web server

In [ ]:
# Install sshtunnel
!pip install sshtunnel --quiet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 227.3/227.3 kB 12.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 284.2/284.2 kB 16.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 856.7/856.7 kB 43.5 MB/s eta 0:00:00


In [ ]:
# Load the SSHTunnelForwarder package
from sshtunnel import SSHTunnelForwarder

In [ ]:
# Define my SSH and MySql connection parameters
ssh_host = "vps31884.dreamhostps.com"
ssh_port = 22  # Default SSH port
ssh_username = "petpuffs"
ssh_password = userdata.get('db_pwd')  # or use ssh_pkey for private key authentication

mysql_host = "mysql.joedag32.com"  # Often "127.0.0.1" when tunneling
mysql_port = 3306  # Default MySQL port
mysql_username = "petpuffs"
mysql_password = userdata.get('db_pwd')

In [ ]:
# Create and start SSH tunnel to create the dog_breed table
with SSHTunnelForwarder(
    (ssh_host, ssh_port),
    ssh_username=ssh_username,
    ssh_password=ssh_password,
    remote_bind_address=(mysql_host, mysql_port),
) as server:
    print(f"SSH tunnel established on port {server.local_bind_port}")
    # MySQL connection must go within the within the with

    # Connect to MySQL using SSH tunnel
    conn = mysql.connector.connect(
            host="127.0.0.1", # Use localhost since we're connected via SSH
            port=server.local_bind_port,
            user=mysql_username,
            password=mysql_password,
            database="data_warehousing_final",
        )

    cursor = conn.cursor()
    #cursor.execute("SELECT VERSION()")
    #data = cursor.fetchone()
    #print(f"Database version: {data[0]}")

    # let's create our dog_breed table
    create_breed_table_query = """CREATE TABLE IF NOT EXISTS dog_breed (
                                  id INT AUTO_INCREMENT PRIMARY KEY,
                                  breed VARCHAR(255),
                                  temperament VARCHAR(255),
                                  popularity FLOAT,
                                  min_height FLOAT,
                                  max_height FLOAT,
                                  min_weight FLOAT,
                                  max_weight FLOAT,
                                  min_expectancy FLOAT,
                                  max_expectancy FLOAT,
                                  energy_level_value FLOAT,
                                  energy_level_category VARCHAR(255),
                                  trainability_value FLOAT,
                                  trainability_category VARCHAR(255),
                                  demeanor_value FLOAT,
                                  demeanor_category VARCHAR(255),
                                  created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                                  updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP
                                  );
                               """
    cursor.execute(create_breed_table_query) # run our query to create the dog_breed table

    conn.close() # We always close the db of course!

SSH tunnel established on port 38539


In [ ]:
# Create and start SSH tunnel to populate the dog_breed table
with SSHTunnelForwarder(
    (ssh_host, ssh_port),
    ssh_username=ssh_username,
    ssh_password=ssh_password,
    remote_bind_address=(mysql_host, mysql_port),
) as server:
    print(f"SSH tunnel established on port {server.local_bind_port}")
    # MySQL connection must go within the within the with

    # Connect to MySQL using SSH tunnel
    conn = mysql.connector.connect(
            host="127.0.0.1", # Use localhost since we're connected via SSH
            port=server.local_bind_port,
            user=mysql_username,
            password=mysql_password,
            database="data_warehousing_final",
        )

    cursor = conn.cursor()

    # for look to go through each row of the akc_dogs_df
    for index, dog in akc_dogs_df.iterrows():

      # INSERT into dog_breed table
      cursor.execute("""INSERT IGNORE INTO dog_breed (
                          breed,
                          temperament,
                          popularity,
                          min_height,
                          max_height,
                          min_weight,
                          max_weight,
                          min_expectancy,
                          max_expectancy,
                          energy_level_value,
                          energy_level_category,
                          trainability_value,
                          trainability_category,
                          demeanor_value,
                          demeanor_category
                          )
                        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                     """, (
                          dog['breed'],
                          dog['temperament'],
                          dog['popularity'],
                          dog['min_height'],
                          dog['max_height'],
                          dog['min_weight'],
                          dog['max_weight'],
                          dog['min_expectancy'],
                          dog['max_expectancy'],
                          dog['energy_level_value'],
                          dog['energy_level_category'],
                          dog['trainability_value'],
                          dog['trainability_category'],
                          dog['demeanor_value'],
                          dog['demeanor_category']
                     ))
      # conn.commit() # Commenting this out so we don't keep adding to the table by accident. Please uncomment if we need to insert to the table again.

    conn.close() # We always close the db of course!

SSH tunnel established on port 45771


We have our dog_breed table created and fully populated on our MySQL web server now!!

In [ ]:
# Create and start SSH tunnel to create the pet and pet_outcome table
with SSHTunnelForwarder(
    (ssh_host, ssh_port),
    ssh_username=ssh_username,
    ssh_password=ssh_password,
    remote_bind_address=(mysql_host, mysql_port),
) as server:
    print(f"SSH tunnel established on port {server.local_bind_port}")
    # MySQL connection must go within the within the with

    # Connect to MySQL using SSH tunnel
    conn = mysql.connector.connect(
            host="127.0.0.1", # Use localhost since we're connected via SSH
            port=server.local_bind_port,
            user=mysql_username,
            password=mysql_password,
            database="data_warehousing_final",
        )

    cursor = conn.cursor()

    # let's create our pet table
    create_pet_table_query = """CREATE TABLE IF NOT EXISTS pet (
                                  id INT AUTO_INCREMENT PRIMARY KEY,
                                  name VARCHAR(255),
                                  dob DATE,
                                  sex VARCHAR(255),
                                  color VARCHAR(255),
                                  age VARCHAR(255),
                                  created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                                  updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
                                  dog_breed_id INT,
                                  FOREIGN KEY (dog_breed_id) REFERENCES dog_breed(id)
                                  );
                               """
    cursor.execute(create_pet_table_query) # run our query to create the dog_breed table

    # let's create our pet_outcome table
    create_pet_outcome_table_query = """CREATE TABLE IF NOT EXISTS pet_outcome (
                                  id INT AUTO_INCREMENT PRIMARY KEY,
                                  outcome_type VARCHAR(255),
                                  outcome_subtype VARCHAR(255),
                                  outcome_datetime DATETIME,
                                  created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                                  updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
                                  pet_id INT,
                                  FOREIGN KEY (pet_id) REFERENCES pet(id)
                                  );
                               """
    cursor.execute(create_pet_outcome_table_query) # run our query to create the dog_breed table

    conn.close() # We always close the db of course!

SSH tunnel established on port 33889


Let's populate the pet and pet_outcome tables. This will be a little tricky with the loops and conditionals, so we'll practice the the logic first.

In [ ]:
# Let's get the dog_breed ids and assign to a lookup dataframe for efficiency
with SSHTunnelForwarder(
    (ssh_host, ssh_port),
    ssh_username=ssh_username,
    ssh_password=ssh_password,
    remote_bind_address=(mysql_host, mysql_port),
) as server:
    print(f"SSH tunnel established on port {server.local_bind_port}")
    # MySQL connection must go within the with

    # Connect to MySQL using SSH tunnel
    conn = mysql.connector.connect(
            host="127.0.0.1", # Use localhost since we're connected via SSH
            port=server.local_bind_port,
            user=mysql_username,
            password=mysql_password,
            database="data_warehousing_final",
        )

    cursor = conn.cursor()

    # Run the SQL to get the id of the breed ids and breed
    cursor.execute("""SELECT id, breed
                    FROM dog_breed
                    """)
    breed_lookup_dict = {}
    for breed in cursor.fetchall():
      breed_lookup_dict[breed[1]] = breed[0]

    conn.commit() # Commit the changes after the loop
    conn.close() # We always close the db of course!

    print(breed_lookup_dict)

SSH tunnel established on port 37045
{'Affenpinscher': 1, 'Afghan Hound': 2, 'Airedale Terrier': 3, 'Akita': 4, 'Alaskan Malamute': 5, 'American Bulldog': 6, 'American English Coonhound': 7, 'American Eskimo Dog': 8, 'American Foxhound': 9, 'American Hairless Terrier': 10, 'American Leopard Hound': 11, 'American Staffordshire Terrier': 12, 'American Water Spaniel': 13, 'Anatolian Shepherd Dog': 14, 'Appenzeller Sennenhund': 15, 'Australian Cattle Dog': 16, 'Australian Kelpie': 17, 'Australian Shepherd': 18, 'Australian Stumpy Tail Cattle Dog': 19, 'Australian Terrier': 20, 'Azawakh': 21, 'Barbet': 22, 'Basenji': 23, 'Basset Fauve de Bretagne': 24, 'Basset Hound': 25, 'Bavarian Mountain Scent Hound': 26, 'Beagle': 27, 'Bearded Collie': 28, 'Beauceron': 29, 'Bedlington Terrier': 30, 'Belgian Laekenois': 31, 'Belgian Malinois': 32, 'Belgian Sheepdog': 33, 'Belgian Tervuren': 34, 'Bergamasco Sheepdog': 35, 'Berger Picard': 36, 'Bernese Mountain Dog': 37, 'Bichon Frise': 38, 'Biewer Terrier

In [ ]:
# # Comment out this block. It was a very inefficient first attempt a the pet and pet_outcome INSERT
# # Create and start SSH tunnel to populate the pet and pet_outcome tables
# with SSHTunnelForwarder(
#     (ssh_host, ssh_port),
#     ssh_username=ssh_username,
#     ssh_password=ssh_password,
#     remote_bind_address=(mysql_host, mysql_port),
# ) as server:
#     print(f"SSH tunnel established on port {server.local_bind_port}")
#     # MySQL connection must go within the with

#     # Connect to MySQL using SSH tunnel
#     conn = mysql.connector.connect(
#             host="127.0.0.1", # Use localhost since we're connected via SSH
#             port=server.local_bind_port,
#             user=mysql_username,
#             password=mysql_password,
#             database="data_warehousing_final",
#         )

#     cursor = conn.cursor()

#     # Test the loop and conditional logic for the INSERTS
#     for index, pet in filtered_pet_trim_df.iterrows():
#         current_breed = pet['breed'] # set the current breed to current_breed
#         breed_id = breed_lookup_dict[current_breed] # get the breed_id from the breed_lookup_dict

#         # Run the INSERT into pet table
#         cursor.execute("""INSERT IGNORE INTO pet (
#                             name,
#                             dob,
#                             sex,
#                             color,
#                             age,
#                             dog_breed_id
#                             )
#                           VALUES (%s, %s, %s, %s, %s, %s)
#                        """, (
#                             pet['name'],
#                             pet['dob'],
#                             pet['sex'],
#                             pet['color'],
#                             pet['age'],
#                             breed_id
#                       ))

#         # Get the last inserted pet ID
#         pet_id = cursor.lastrowid

#         # If pet_outcome is not equal to None, then insert a record into the pet_outcome table
#         if pet['outcome_type'] != 'None':
#             cursor.execute("""INSERT IGNORE INTO pet_outcome (
#                                 outcome_type,
#                                 outcome_subtype,
#                                 outcome_datetime,
#                                 pet_id
#                                 )
#                               VALUES (%s, %s, %s, %s)
#                            """, (
#                                 pet['outcome_type'],
#                                 pet['outcome_subtype'],
#                                 pet['outcome_datetime'],
#                                 pet_id
#                            ))
#     #
#     # Commenting this out, this approach was very inefficient and testing out a batch below !!!
#     # conn.commit() # Commit the changes after the loop
#     conn.close() # We always close the db of course!

In [ ]:
# let's try a batch insert query for better efficiency
# Create and start SSH tunnel to populate the pet and pet_outcome tables
with SSHTunnelForwarder(
    (ssh_host, ssh_port),
    ssh_username=ssh_username,
    ssh_password=ssh_password,
    remote_bind_address=(mysql_host, mysql_port),
) as server:
    print(f"SSH tunnel established on port {server.local_bind_port}")

    # MySQL connection must go within the with block
    conn = mysql.connector.connect(
            host="127.0.0.1",  # Use localhost for SSH tunnel
            port=server.local_bind_port,
            user=mysql_username,
            password=mysql_password,
            database="data_warehousing_final",
        )

    cursor = conn.cursor()

    # Prepare batch insert queries
    pet_insert_query = """
        INSERT IGNORE INTO pet (name, dob, sex, color, age, dog_breed_id)
        VALUES (%s, %s, %s, %s, %s, %s)
    """
    pet_outcome_insert_query = """
        INSERT IGNORE INTO pet_outcome (outcome_type, outcome_subtype, outcome_datetime, pet_id)
        VALUES (%s, %s, %s, %s)
    """

    # Collect data for batch inserts in pet_data and pet_outcome_data arrays
    pet_data = []
    pet_outcome_data = []
    # loop through the filtered_pet_trim_df dataframe
    for index, pet in filtered_pet_trim_df.iterrows():
        current_breed = pet['breed'] # Assign the current row breed to current_breed
        breed_id = breed_lookup_dict[current_breed] # lookup the matching breed in the breed_lookup_dict and assign its number to breed_id

        # add the current loop row values to the pet_data incluing our breed_id we just built above
        pet_data.append((pet['name'], pet['dob'], pet['sex'], pet['color'], pet['age'], breed_id))

        if pet['outcome_type'] != 'None':
            pet_outcome_data.append((pet['outcome_type'], pet['outcome_subtype'], pet['outcome_datetime'], index + 1))  # Assuming pet IDs start from 1

    # Execute batch inserts
    cursor.executemany(pet_insert_query, pet_data)
    cursor.executemany(pet_outcome_insert_query, pet_outcome_data)

    # Comment out the conn.commit so that we don't re-populate our tables. Uncomment if we need to run the query again.
    # conn.commit()  # Commit changes
    conn.close()  # Close the connection

SSH tunnel established on port 33999


The above query using a batch insert and breed_lookup_dict to lookup breed id's literally turned a 4+ hour process into a 7-8 second process. Amazing!!! That would be huge cost savings if we ran that often on a cloud platform!

## Visualizations

In [ ]:
# Import the seaborn and maytplotlib libraries to build a graph
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
# Let's select the top 10 most energetic breeds according AKC from dog_breed
with SSHTunnelForwarder(
    (ssh_host, ssh_port),
    ssh_username=ssh_username,
    ssh_password=ssh_password,
    remote_bind_address=(mysql_host, mysql_port),
) as server:
    print(f"SSH tunnel established on port {server.local_bind_port}")
    # MySQL connection must go within the with

    # Connect to MySQL using SSH tunnel
    conn = mysql.connector.connect(
            host="127.0.0.1", # Use localhost since we're connected via SSH
            port=server.local_bind_port,
            user=mysql_username,
            password=mysql_password,
            database="data_warehousing_final",
        )


    #Use pandas to read the sql query from the db
    energetic_breeds_query = pd.read_sql("""SELECT breed, energy_level_value
                                           FROM dog_breed
                                           WHERE energy_level_value > 0
                                           ORDER BY energy_level_value DESC
                                           LIMIT 10
                                         """, conn)

    conn.close() # We always close the db of course!

energetic_breeds_query # Test the results

SSH tunnel established on port 44093


  energetic_breeds_query = pd.read_sql("""SELECT breed, energy_level_value



breed  energy_level_value
0                 Beauceron                 1.0
1             Border Collie                 1.0
2     Australian Cattle Dog                 1.0
3     Catahoula Leopard Dog                 1.0
4  Braque Francais Pyrenean                 1.0
5     Czechoslovakian Vlcak                 1.0
6             Berger Picard                 1.0
7         Croatian Sheepdog                 1.0
8                  Brittany                 1.0
9    Appenzeller Sennenhund                 1.0